cpp/src/binaryop/compiled/binops_custom.cuh

```diff
@@ -0,0 +1,160 @@
+  __device__ void operator()(Args...)
+  {
+  }
+};
+
+/**
+ * @brief Functor to launch only defined operations without common type.
+ *
+ * @tparam BinaryOperator binary operator functor
+ */
+template <typename BinaryOperator>
+struct ops2_wrapper {
+  mutable_column_device_view& out;
+  column_device_view const& lhs;
+  column_device_view const& rhs;
+  template <typename TypeLhs,
+            typename TypeRhs,
+            std::enable_if_t<!has_common_type_v<TypeLhs, TypeRhs> and
+                             is_op_supported<TypeLhs, TypeRhs, BinaryOperator>()>* = nullptr>
+  __device__ void operator()(size_type i)
+  {
+    TypeLhs x   = lhs.element<TypeLhs>(i);
+    TypeRhs y   = rhs.element<TypeRhs>(i);
+    auto result = BinaryOperator{}.template operator()<TypeLhs, TypeRhs>(x, y);
+    type_dispatcher(out.type(), typed_casted_writer<decltype(result)>{}, i, out, result);
+  }
+
+  template <typename TypeLhs,
+            typename TypeRhs,
+            typename... Args,
+            std::enable_if_t<has_common_type_v<TypeLhs, TypeRhs> or
+                             not is_op_supported<TypeLhs, TypeRhs, BinaryOperator>()>* = nullptr>
+  __device__ void operator()(Args...)
+  {
+  }
+};
+
+/**
+ * @brief Functor which does single, and double type dispatcher in device code
+ *
+ * single type dispatcher for lhs and rhs with common types.
+ * double type dispatcher for lhs and rhs without common types.
+ *
+ * @tparam BinaryOperator binary operator functor
+ */
+template <class BinaryOperator>
+struct device_type_dispatcher {
+  //, OperatorType type)
+  // (type == OperatorType::Direct ? operator_name : 'R' + operator_name);
+  mutable_column_device_view out;
+  column_device_view lhs;
+  column_device_view rhs;
+  data_type common_data_type;
+  device_type_dispatcher(mutable_column_device_view ot,
+                         column_device_view lt,
+                         column_device_view rt,
+                         data_type ct)
+    : out(ot), lhs(lt), rhs(rt), common_data_type(ct)
+  {
+  }
+
+  __device__ void operator()(size_type i)
+  {
+    if (common_data_type == data_type{type_id::EMPTY}) {
+      double_type_dispatcher(
+        lhs.type(), rhs.type(), ops2_wrapper<BinaryOperator>{out, lhs, rhs}, i);
+    } else {
+      type_dispatcher(common_data_type, ops_wrapper<BinaryOperator>{out, lhs, rhs}, i);
+    }
+  }
+};
+}  // namespace
+/**
+ * @brief Functor that runs binary operation on each element of @p lhsd and @p rhsd columns.
+ *
+ * This template is instantiated for each binary operator.
+ *
+ * @tparam BinaryOperator Binary operator functor
+ * @param outd mutable device view of output column
+ * @param lhsd device view of left operand column
+ * @param rhsd device view of right operand column
+ * @param stream CUDA stream used for device memory operations
+ */
+template <class BinaryOperator>
+void dispatch_single_double(mutable_column_device_view& outd,
```

Comment:
If you move the definition of this template to it's own translation unit (i.e., the definition is hidden in TUs where `binops_custom.cuh` is included) then you don't need all of the `extern` declarations. It will find your explicit template instantiations later. See https://stackoverflow.com/a/59614755/11341974

Reply:
done.